internal/handlers/common/findandmodify.go

```diff
@@ -168,3 +178,152 @@ func GetFindAndModifyParams(doc *types.Document, l *zap.Logger) (*FindAndModifyP
 		MaxTimeMS:          maxTimeMS,
 	}, nil
 }
+
+// UpsertDocument updates the first document if exists, or create an insert document
+// from the params if no documents in query result docs.
+func UpsertDocument(docs []*types.Document, params *FindAndModifyParams) (*UpsertParams, error) {
+	res := new(UpsertParams)
+	var err error
+
+	if len(docs) == 0 {
+		res.Insert, err = insertDocuments(params)
+
+		res.ReturnValue = types.Null
+		if params.ReturnNewDocument {
+			res.ReturnValue = res.Insert
+		}
+
+		return res, err
+	}
+
+	res.Update, err = updateDocuments(docs, params)
+
+	res.ReturnValue = docs[0]
+	if params.ReturnNewDocument {
+		res.ReturnValue = res.Update
+	}
+
+	return res, err
+}
+
+// insertDocuments creates an insert document from the parameter.
+// When inserting new document we must check that `_id` is present, so we must extract `_id`
+// from query or generate a new one.
+func insertDocuments(params *FindAndModifyParams) (*types.Document, error) {
```

Comment:
It's a bit confusing to call it `insertDocuments` because, as far as I can see, this function doesn't insert anything. Maybe it could be called something like `prepareDocumentForInsert`? It prepares a single document, right?